lang/c/src/value-read-resolution.c

```diff
@@ -0,0 +1,628 @@
+    size_t  i;
+    
+    avro_schema_t  record_schema_reader = avro_value_get_schema(dest);
+    avro_schema_t  record_schema_writer = avro_value_get_schema(source);
+    
+    size_t reader_fields_count;
+    size_t writers_fields_count;
+    
+    check(rval, avro_value_get_size(dest, &reader_fields_count));
+    check(rval, avro_value_get_size(source, &writers_fields_count));
+    
+    // map to keep track of already resolved fields.
+    st_table* reader_fields_table = st_init_strtable();
+    
+    // loop writers fields
+    for(i = 0; i<writers_fields_count; ++i)
+    {
+        avro_value_t field;
+        check(rval, avro_value_get_by_index(source, i, &field, NULL));
+        char* field_name =  avro_schema_record_field_name(record_schema_writer, i);
+        
+        int reades_index = avro_schema_record_field_get_index(record_schema_reader, field_name);
+        
+        if (reades_index > -1) {
+            // ok, reader's schema has the field with the same name too,
+            // lets do schema resolution.
+            avro_value_t reader_field;
+            check(rval, avro_value_get_by_index(dest, reades_index, &reader_field, NULL));
+            
+            check(rval, read_value_with_resolution(reader, &field, &reader_field));
+
+            st_insert(reader_fields_table, (st_data_t) field_name, &reader_field);
+            
+        } else {
+            // reader's schema doesn't have this field, skip it.
+            avro_schema_t  field_schema = avro_schema_record_field_get_by_index(record_schema_writer, i);
+            check(rval, avro_skip_data(reader, field_schema));
+        }
+    }
+    
+    // Check remaining reader's schema fields for default values.
+    for(i = 0; i<reader_fields_count; ++i)
+    {
+        avro_value_t field;
+        check(rval, avro_value_get_by_index(dest, i, &field, NULL));
+        char* readers_field_name = avro_schema_record_field_name(record_schema_reader, i);
+        
+        st_data_t  data;
+        if(st_lookup(reader_fields_table, (st_data_t) readers_field_name, &data) == 0)
+        {
+            // this field is not in writer's schema, it must have default value.
+            json_t* def_val = avro_schema_record_field_default(record_schema_reader, readers_field_name);
+            int has_def = def_val == NULL ? 0 : 1;
+            if(has_def == 0)
+            {
+                avro_set_error("Schema resolution error! There must be default "
+                               "value set for fields missing in writer's schema.");
+                return EINVAL;
+            }
+            // read default value
+            read_default_value(record_schema_reader, readers_field_name, &field);
+        }
+    }
+
+    if(reader_fields_table)
+    {
+        // cleanup
+        st_free_table(reader_fields_table);
+    }
+    return 0;
+}
+
+// Simple macro to check if dest field is union before setting the value.
+#define avro_resolve_source(reader, source, dest, type) \
```

Comment:
Macros are pure evil. But if you insist, I think it's good practice to use `ALL_CAPS` for their names, so they aren't masquerading to the casual observer as a function call, which they are not.